### PR TITLE
Refactoring with Lombok

### PR DESCRIPTION
**Pull request by bonaparten**
For further discussions: #50

### CODE REVIEW

File: src/main/java/org/heigit/ohsome/ohsomeapi/output/dataaggregationresponse/groupbyresponse/GroupByObject.java [hunk 1]
```diff
@@ -11,16 +13,10 @@
...
-  protected Object groupByObjectValue;
-
-  public GroupByObject(Object groupByObjectValue) {
-    this.groupByObjectValue = groupByObjectValue;
-  }
-
-  public Object getGroupByObject() {
-    return groupByObjectValue;
-  }
+  protected Object groupByObject;
```
Comment (FabiKo117) [inline src/main/java/org/heigit/ohsome/ohsomeapi/output/dataaggregationresponse/groupbyresponse/GroupByObject.java:21]: Why did you rename this variable?

File: src/main/java/org/heigit/ohsome/ohsomeapi/output/dataaggregationresponse/DefaultAggregationResponse.java [hunk 2]
```diff
@@ -4,15 +4,18 @@
...
  * Represents the outer JSON response object for the data aggregation requests that do not use the
  * /groupBy resource. It contains attribution info, the version of the api, optional
- * {@link org.heigit.ohsome.ohsomeapi.output.dataaggregationresponse.Metadata
- * Metadata} and the
+ * {@link org.heigit.ohsome.ohsomeapi.output.dataaggregationresponse.Metadata Metadata} and the
  * {@link org.heigit.ohsome.ohsomeapi.output.dataaggregationresponse.elements.ElementsResult
  * ElementsResult} objects.
  */
+@Getter
+@NoArgsConstructor
```
Comment (FabiKo117) [inline src/main/java/org/heigit/ohsome/ohsomeapi/output/dataaggregationresponse/DefaultAggregationResponse.java:18]: I think here you'd also have to define _access = AccessLevel.PRIVATE_ right?

**Review by FabiKo117 — CHANGES_REQUESTED** [review]
see inline comments

File: src/main/java/org/heigit/ohsome/ohsomeapi/output/dataaggregationresponse/groupbyresponse/GroupByObject.java (hunk 1 above)
Comment (bonaparten) [reply src/main/java/org/heigit/ohsome/ohsomeapi/output/dataaggregationresponse/groupbyresponse/GroupByObject.java:21]: I renamed it because Lombok search for "groubByObjectValue" in the json response, which use although "groupByObject". We get a NullpointerException since "groupByObjectValue" is not present in the json response.
The problem with this change is that the variable name is now the same as the class name.

File: src/main/java/org/heigit/ohsome/ohsomeapi/output/dataaggregationresponse/DefaultAggregationResponse.java (hunk 2 above)
Comment (bonaparten) [reply src/main/java/org/heigit/ohsome/ohsomeapi/output/dataaggregationresponse/DefaultAggregationResponse.java:18]: yes, you're right

File: src/main/java/org/heigit/ohsome/ohsomeapi/inputprocessing/InputProcessor.java [hunk 3]
```diff
@@ -100,10 +104,6 @@ public InputProcessor(HttpServletRequest servletRequest, boolean isSnapshot, boo
     this.requestParameters = servletRequest.getParameterMap();
   }
 
-  public InputProcessor(ProcessingData processingData) {
```
[the remark this answers is not in the document]
Comment (bonaparten) [reply src/main/java/org/heigit/ohsome/ohsomeapi/inputprocessing/InputProcessor.java:103]: If we want to use Lombok, I think we should establish in which cases Lombok has to be used and in which cases avoided.
This is one of more cases in which the use of Lombok is possible but not necessary convenient.

[the remark this answers is not in the document]
Comment (FabiKo117) [reply src/main/java/org/heigit/ohsome/ohsomeapi/inputprocessing/InputProcessor.java:103]: In my opinion, the other usage examples here in this branch are totally fitting, e.g. hiding the standard getter/setter writing, as well as the "standard constructor usage" meaning cases where you would use ```@AllArgsConstructor```.

But in cases like here, when we have more than one constructor and it also includes some processing, I'd rather avoid it.
Curious what @rtroilo or @tyrasd would have to say about that too.

File: src/main/java/org/heigit/ohsome/ohsomeapi/output/dataaggregationresponse/groupbyresponse/GroupByObject.java (hunk 1 above)
Comment (FabiKo117) [reply src/main/java/org/heigit/ohsome/ohsomeapi/output/dataaggregationresponse/groupbyresponse/GroupByObject.java:21]: I see.. yeah that's one problem. Another one is that the variable name now does not reflect for what this variable actually stands, as this is it's not an object, but just an identifier, e.g. "Heidelberg". But yeah changing it to something like ```groupByObjectId``` would cause a change in the response as well. I think this would still be an improvement though, so I'd suggest to call it then ```groupByObjectId```.

File: src/main/java/org/heigit/ohsome/ohsomeapi/inputprocessing/InputProcessor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (bonaparten) [reply src/main/java/org/heigit/ohsome/ohsomeapi/inputprocessing/InputProcessor.java:103]: There are another cases which could be a little bit misleading, for example [here](https://github.com/GIScience/ohsome-api/blob/64b3d53e58cf4d368625811528fa115f5b703b2f/src/main/java/org/heigit/ohsome/ohsomeapi/executor/ExecutionUtils.java#L94)
It is possible to use only private and not-initialized variables with the annotation @RequiredArgsConstructor. Since the constructor here takes only `ProcessingData processingData;`, which is also the only variable of this kind in the class, this annotation works for the purpose but it could mislead the developer.

File: src/main/java/org/heigit/ohsome/ohsomeapi/output/dataaggregationresponse/groupbyresponse/GroupByObject.java (hunk 1 above)
Comment (bonaparten) [reply src/main/java/org/heigit/ohsome/ohsomeapi/output/dataaggregationresponse/groupbyresponse/GroupByObject.java:21]: Nice, I will change the variable name.

**Review by tyrasd — COMMENTED** [review]
I found some formatting/checkstyle issues.

regarding the main "content" of the PR, I'm missing a bit the rationale behind these changes. I see the amount of "boring" code like getter/setters is reduced. But it comes (at least in my IDE IntelliJ) at the cost that helpful stuff like code autocompletion doesn't work anymore. Or is there a trick to get this integrated into an IDE which I don't see right now, e.g., does one need to install a plugin or something?

**Comment by bonaparten** [comment]
> I found some formatting/checkstyle issues.
> 
> regarding the main "content" of the PR, I'm missing a bit the rationale behind these changes. I see the amount of "boring" code like getter/setters is reduced. But it comes (at least in my IDE IntelliJ) at the cost that helpful stuff like code autocompletion doesn't work anymore. Or is there a trick to get this integrated into an IDE which I don't see right now, e.g., does one need to install a plugin or something?
> 
> //edit: I tried the instructions from https://www.baeldung.com/lombok-ide, but didn't have much luck. disappointed
> //edit2: ok, after a second try it worked (I must have misclicked something the first time). anyway, this really needs to be documented somewhere (readme, contrib. guide, …) IMO.

I  can write something on confluence about it or elsewhere. Where is it better? @FabiKo117 @tyrasd

**Comment by tyrasd** [comment]
IMHO, for now https://github.com/GIScience/ohsome-api#getting-started seems like the best place to start. Or alternatively start a github wiki page.

File: src/test/java/org/heigit/ohsome/ohsomeapi/controller/GetControllerTest.java [hunk 4]
```diff
@@ -158,13 +159,15 @@ public void getElementsCountGroupByBoundaryGroupByTagTest() {
...
         JsonNode.class);
-    assertEquals(43, StreamSupport
-        .stream(Spliterators.spliteratorUnknownSize(
-            response.getBody().get("groupByResult").iterator(), Spliterator.ORDERED), false)
-        .filter(
-            jsonNode -> "boundary1".equalsIgnoreCase(jsonNode.get("groupByObject").get(0).asText())
-                && "remainder".equalsIgnoreCase(jsonNode.get("groupByObject").get(1).asText()))
-        .findFirst().get().get("result").get(0).get("value").asInt(), 0);
+    assertEquals(43,
+        StreamSupport
```
[the remark this answers is not in the document]
Comment (bonaparten) [reply src/test/java/org/heigit/ohsome/ohsomeapi/controller/GetControllerTest.java:163]: I have no answer for that, it is automatically formatted. I'm not using specific format settings which regard this issue.

[the remark this answers is not in the document]
Comment (tyrasd) [reply src/test/java/org/heigit/ohsome/ohsomeapi/controller/GetControllerTest.java:163]: ok, I see. If you ask me, this is not a good situation to use automatic code formatting, for multiple reasons:

* First, the code was already correctly formatted, so why change it?
* Since we're not all using the same IDEs, if another person uses a different auto code formatter, the same line would potentially be changed again. This could result in an "edit war" of the auto formatters.
* Such changes make it hard to read and review a pull request, because more lines are touched and it is sometimes hard to see whether something really changed or if it was "just" the reordering of a few newlines and whitespaces.
* Most importantly, any such line which is touched without a reason can lead to unnecessary merge conflicts with other branches. This can cause lots of manual work down the line.

You might choose to use an auto-formatter during your own development however you like, but if you do, please make sure not to commit unnecessary changes which were just introduced automatically. [`git add -p` is your friend](https://johnkary.net/blog/git-add-p-the-most-powerful-git-feature-youre-not-using-yet/).

[the remark this answers is not in the document]
Comment (FabiKo117) [reply src/test/java/org/heigit/ohsome/ohsomeapi/controller/GetControllerTest.java:163]: I also ran into that issue in the past (and sometimes sitll today), which can indeed be a bit annoying and missleading.

What I try to remember myself always now is just marking the specific code lines, or method, that I've added/edited and then apply the auto formatting on that, instead of the whole class.
Sometimes it really changes something on a completely other part of the code, which also fits to the coding guidelines, but is just another way of ordering the lines..

File: src/main/java/org/heigit/ohsome/ohsomeapi/executor/ExecutionUtils.java [hunk 5]
```diff
@@ -87,17 +87,15 @@
 import org.locationtech.jts.geom.Polygonal;
 import org.locationtech.jts.geom.Puntal;
 import org.wololo.jts2geojson.GeoJSONWriter;
+import lombok.RequiredArgsConstructor;
```
[the remark this answers is not in the document]
Comment (FabiKo117) [reply src/main/java/org/heigit/ohsome/ohsomeapi/executor/ExecutionUtils.java]: To set the alphabetical sorting in Eclipse, you need to do what is suggested here in the last answer with screenshots:
https://stackoverflow.com/questions/34450900/how-to-sort-import-statements-in-eclipse-in-case-insensitive-order

**Review by FabiKo117 — APPROVED** [review]
looks good now 👍

**Comment by tyrasd** [comment]
360508a6be35a4b0c2d6a9e731d911f1840d3d9e changed the JSON output of all `groupBy` requests from `groupByObject` to `groupByObjectId`. this must have slipped through in the code review. :see_no_evil:
